velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -154,18 +166,34 @@ core::PlanNodePtr SubstraitVeloxPlanConverter::toVeloxPlan(
 
   const auto& inputType = childNode->outputType();
   int colIdx = 0;
+  // Note that Substrait projection adds the project expressions on top of the
+  // input to the projection node. Thus we need to add the input columns first
+  // and then add the projection expressions.
+
+  // adding input node columns
+  for (uint32_t idx = 0; idx < inputType->size(); idx++) {
+    const auto& fieldName = inputType->nameOf(idx);
+    projectNames.emplace_back(fieldName);
+    expressions.emplace_back(std::make_shared<core::FieldAccessTypedExpr>(
+        inputType->childAt(idx), fieldName));
+    colIdx += 1;
+  }
+
+  // adding projection columns
   for (const auto& expr : projectExprs) {
     expressions.emplace_back(exprConverter_->toVeloxExpr(expr, inputType));
     projectNames.emplace_back(
         substraitParser_->makeNodeName(planNodeId_, colIdx));
     colIdx += 1;
   }
 
-  return std::make_shared<core::ProjectNode>(
-      nextPlanNodeId(),
-      std::move(projectNames),
-      std::move(expressions),
-      childNode);
+  return ProcessEmit(
```

Comment:
It doesn't makes sense to put emit on top of project. Can we just configure Project to produce columns in the right order?

Reply:
Thinking about the Substrait definition. Let's see how this is mapped in the current definition. 

[RelCommon](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L16-L22) is the entity which contains common attributes for each relation in Substrait. This holds two things, `Direct` and `Emit`. `Direct` output as is, and `Emit` allows to control the order of output also we could omit things from here as well. 

Now let's take a look at [Project](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L148-L153). Like this, each relation has the `RelCommon` as an attribute. So if we ignore emit here, we are ignoring the Substrait definition on Project. Need to think about this a bit, before choosing a different path to get the same outcome.

But I think I understand your point (I assume, please correct me if I am wrong). In Velox, the project definition means we only output what is being defined as project expressions. But if we adopt this, the output will have input fields + project expressions. But we would expect it to only output the project columns only. 

I think in the [VeloxToSubstraitPlan](https://github.com/facebookincubator/velox/blob/bbb2c495a06fc01f2f245e1a221ba518504b2cf1/velox/substrait/VeloxToSubstraitPlan.cpp#L231) some how this has been considered and it adds the emit output mapping to only include the project fields since they are after the `inputs` and thus `inputTypeSize + i` where `i` is the index of the output field.

What do you think?

Reply:
My understanding is that Substrait defined Project node by (1) specifying a list of expressions; (2) output order.

If output order is direct, then project node is expected to return all inputs columns + new columns computed using specified expressions.

If output order is emit, then project node is expected to return a subset of input columns + new columns in the specified order.

Both of these scenarios can be implemented using single ProjectNode in Velox. There is no need to create 2 ProjectNodes one after another.

Reply:
I updated the PR.